include/frozen/bits/pic_array.h

```diff
@@ -0,0 +1,549 @@
+  constexpr       iterator end()         noexcept { return {array_data_, std::end(data_)  }; }
+  constexpr const_iterator end()   const noexcept { return {array_data_, std::end(data_)  }; }
+
+  // Capacity
+  constexpr size_type size() const noexcept { return N; }
+  constexpr size_type max_size() const noexcept { return N; }
+
+  // Element access
+  constexpr       reference operator[](std::size_t index)       { return begin()[index]; }
+  constexpr const_reference operator[](std::size_t index) const { return begin()[index]; }
+
+private:
+  friend iterator;
+  friend const_iterator;
+
+  // Intermediate functor, required because value_type might not be constexpr default constructible
+  // or constexpr copy assignable. This way it only needs to be constexpr copy or move assignable.
+  struct appender {
+    size_type store_pos = 0;
+    element_t<value_type> array_data[storage_size] = {};
+
+    template <typename View>
+    constexpr auto operator()(const View& item)
+      -> std::enable_if_t<
+        std::is_same<element_t<View>, element_t<value_type>>::value
+      , ext_array<typename View::value_type, size_type>
+      >
+    {
+      const bool is_nul_terminated = !item.empty() && item.back() == '\0';
+
+      pic_store_type<View, size_type> str_value{store_pos, static_cast<size_type>(item.size())};
+      constexpr_assert(str_value.data_size == item.size(), "size field causes truncation");
+
+      // constexpr std::copy before C++20
+      for (const auto& val : item) {
+        array_data[store_pos++] = val;
+      }
+      if (is_nul_terminated)
+        str_value.data_size -= 1;
+      else
+        store_pos += 1; // NUL-terminator
+
+      return str_value;
+    }
+
+    template <typename ViewL, typename ViewR>
+    constexpr auto operator()(const std::pair<ViewL, ViewR>& item)
+      -> std::enable_if_t<
+        std::is_same<element_t<ViewL>, element_t<value_type>>::value
+        && std::is_same<element_t<ViewL>, element_t<ViewR>>::value
+      , storage_type
+      >
+    {
+      return storage_type{(*this)(item.first), (*this)(item.second)};
+    }
+
+    template <typename View, typename U>
+    constexpr auto operator()(const std::pair<View, U>& item)
+      -> std::enable_if_t<
+        std::is_same<element_t<View>, element_t<value_type>>::value
+        && !has_type<element_type<remove_cv_t<U>>>::value
+      , storage_type
+      >
+    {
+      return storage_type{(*this)(item.first), item.second};
+    }
+
+    template <typename View, typename U>
+    constexpr auto operator()(const std::pair<U, View>& item)
+      -> std::enable_if_t<
+        std::is_same<element_t<View>, element_t<value_type>>::value
+        && !has_type<element_type<remove_cv_t<U>>>::value
+      , storage_type
+      >
+    {
+      return storage_type{item.first, (*this)(item.second)};
+    }
+  };
+};
+
+} // namespace bits
+
+// Helpers to preserve arrays in type information, instead of letting them decay to pointers
```

Comment:
This is the primary thing I'm not super happy with. But I couldn't find any other way to preserve the array size (string size) in type information otherwise.

It cannot be put in an argument to a `constexpr` function either, because the *values* of parameters to `constexpr` functions must also be usable at runtime, and thus cannot become part of the type info later on (i.e. `storage_size` template parameter of return type of factory function `make_XXX`).